proposals/roles.md

```diff
@@ -0,0 +1,155 @@
+
+### Member access 
+
+We modify the [member access rules](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1176-member-access) as follows:
+
+- ... the result is that namespace.
+- ... the result is that type constructed with the given type arguments.
+- If `E` is classified as a type, if `E` is not a type parameter, and if a member lookup of `I` in `E` with `K` type parameters produces a match, then `E.I` is evaluated and classified as follows:  
+  > *Note*: When the result of such a member lookup is a method group and `K` is zero, the method group can contain methods having type parameters. This allows such methods to be considered for type argument inferencing. *end note*
+  - If `I` identifies a type, then the result is that type constructed with any given type arguments.
+  - If `I` identifies one or more methods, then the result is a method group with no associated instance expression.
+  - If `I` identifies a static property, then the result is a property access with no associated instance expression.
+  - If `I` identifies a static field:
+    - If the field is readonly and the reference occurs outside the static constructor of the class or struct in which the field is declared, then the result is a value, namely the value of the static field `I` in `E`.
+    - Otherwise, the result is a variable, namely the static field `I` in `E`.
+  - If `I` identifies a static event:
+    - If the reference occurs within the class or struct in which the event is declared, and the event was declared without *event_accessor_declarations*, then `E.I` is processed exactly as if `I` were a static field.
+    - Otherwise, the result is an event access with no associated instance expression.
+  - If `I` identifies a constant, then the result is a value, namely the value of that constant.
+  - If `I` identifies an enumeration member, then the result is a value, namely the value of that enumeration member.
+  - Otherwise, `E.I` is an invalid member reference, and a compile-time error occurs.
+- **Otherwise, if `E` is classified as a type, if `E` is not a type parameter or an extension type, and if an ***extension member lookup*** of `I` in `E` with `K` type parameters produces a match, then `E.I` is evaluated and classified as follows:**  
+  ...
+- **Otherwise, if `E` is classified as an extension type, and if a member lookup of `I` in underlying type `U` with `K` type parameters produces a match, then `E.I` is evaluated and classified as follows:** 
+  ...
+- If `E` is a property access, indexer access, variable, or value, the type of which is `T`, and a member lookup of `I` in `T` with `K` type arguments produces a match, then `E.I` is evaluated and classified as follows:
+  ...
+- Otherwise, an attempt is made to process `E.I` as an extension method invocation. If this fails, `E.I` is an invalid member reference, and a binding-time error occurs.
+
+### Extension member lookup
```

Comment:
:memo: Maybe the "extension member lookup" should be factored into the normal "member lookup".
So other scenarios in the spec that rely on "member lookup", such as resolution of `Add` in collection initializers ([here](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#117154-collection-initializers)) would benefit.

A "member lookup" would become a "simple member lookup" (current rules for member lookup) followed by either an "extension member lookup" or a simple member lookup in the underlying type.

Reply:
:memo: The process of finding a "compatible substituted extension type" is currently proposed in a similar way that we deal with interface unification (detecting potential duplicate/ambiguous interface implementations). We could simplify this problem by placing restrictions on extension types. 
For example, we could:
- require extension types to be non-generic. 
- require the extension's type parameters to match the type parameters of the underlying type.

An example of issue:
`static extension Evil<T, U> : Dictionary<IDisposable, T> where U : IDisposable`
There is no way to go from an instance of `Dictionary'2` to a substituted `Evil'2`.

Another example of complexity:
```
extension E<T, U> : Dictionary<List<T>, U> // T used in nested position
Dictionary<List<int>, string>
```